contracts/RockPaperScissors.sol

```diff
@@ -0,0 +1,334 @@
+        bytes32 gameKey = createPlayerOneMoveHash(msg.sender, secret, Moves(playerOneMove));
+        Game storage game = games[gameKey];
+
+        require(game.step == Steps.PLAYER_TWO_MOVE, INVALID_STEP_MSG);
+        game.step = Steps.PLAYER_ONE_REVEAL;
+
+        uint256 expiryDate = block.timestamp.add(FORFEIT_WINDOW);
+        game.expiryDate = expiryDate;
+
+        game.playerOneMove = Moves(playerOneMove);
+
+        emit PlayerOneReveals(gameKey, msg.sender, Moves(playerOneMove), expiryDate);
+    }
+
+    function revealPlayerTwo(bytes32 gameKey, bytes32 secret, Moves playerTwoMove) external whenNotPaused {
+
+        require(Moves(playerTwoMove) > Moves.EMPTY && Moves(playerTwoMove) <= Moves.SCISSORS, INVALID_MOVE_MSG);
+        Game storage game = games[gameKey];
+        require(game.step == Steps.PLAYER_ONE_REVEAL, INVALID_STEP_MSG);
+        address playerTwo = game.playerTwo;
+
+        //Validate move
+        bytes32 expectedMoveHash = createPlayerTwoMoveHash(msg.sender, gameKey, secret, Moves(playerTwoMove));
+        require(game.playerTwoMoveHash == expectedMoveHash, HASH_MISMATCH_MSG);
+
+        emit PlayerTwoReveals(gameKey, msg.sender, Moves(playerTwoMove));
+
+        uint256 stake = game.stake;
+        address playerOne = game.playerOne;
+        address winner;
+        address loser;
+        GameResult outcome = resolveGame(uint8(game.playerOneMove), uint8(playerTwoMove));
+
+        if (outcome == GameResult.DRAW) {
+            resetGame(game);
+            balances[playerOne] = balances[playerOne].add(stake);
+            balances[playerTwo] = balances[playerTwo].add(stake);
+            DrawRefund(gameKey, playerOne, playerTwo, stake, stake);
+
+            return;
+
+        } else if (outcome == GameResult.LEFTWIN) {
+            winner = playerOne;
+            loser = playerTwo;
+        } else if (outcome == GameResult.RIGHTWIN) {
+            winner = playerTwo;
+            loser = playerOne;
+        } else {
+            require(outcome != GameResult.INCORRECT, "Bugger");
+        }
+        uint256 winnings = stake.mul(2);
+        balances[winner] = balances[winner].add(winnings);
+        emit GamePayment(gameKey, winner, winnings);
+
+        resetGame(game);
+    }
+
+    // Type hinted uint8 because GameResult.INCORRECT will never be returned if Moves enum is used
+    // If the value is out of bounds then an exception will be thrown
+    function resolveGame(uint8 leftPlayer, uint8 rightPlayer) public pure returns(GameResult) {
+
+        if (leftPlayer < uint8(Moves.ROCK) || leftPlayer > uint8(Moves.SCISSORS)) {
+            return GameResult.INCORRECT;
+        }
+
+        if (rightPlayer < uint8(Moves.ROCK) || rightPlayer > uint8(Moves.SCISSORS)) {
+            return GameResult.INCORRECT;
+        }
+
+        return GameResult((uint256(leftPlayer).add(3).sub(uint256(rightPlayer))).mod(3));
```

Comment:
Is the casting to `uint256` necessary?

Reply:
This was to get it to be compatible with SafeMath. I was getting:

```
TypeError: Member "add" not found or not visible after argument-dependent lookup in enum RockPaperScissors.Moves.
        return GameResult((leftPlayer.add(3).sub(rightPlayer)).mod(3));
```

Reply:
A right. I am wondering about a cheeky I have never seen before:

```solidity
using SafeMath for Moves;
```

Reply:
`using SafeMath for Moves;` works! That never occurred to me.

Reply:
Never did it to me until 2 days ago :o

Reply:
Sorry. I think I spoke too soon. Perhaps a delay with my IDE autosave. Ran the compile again and I get:

`TypeError: Member "add" not found or not visible after argument-dependent lookup in enum RockPaperScissors.Moves.`

I had to go back to casting but using uint8() instead of uint256():

`return GameResult((uint8(leftPlayer).add(3).sub(uint8(rightPlayer))).mod(3));`

Reply:
Oh well.